src/watchdog.js

```diff
@@ -0,0 +1,414 @@
+				'or `Watchdog.for` helper.',
+				null
+			);
+		}
+
+		if ( !this._destructor ) {
+			/**
+			 * @error watchdog-destructor-not-defined
+			 *
+			 * The watchdog destructor is not defined, define it using `watchdog.setDestructor()` or `Watchdog.for` helper.
+			 */
+			throw new CKEditorError(
+				'watchdog-destructor-not-defined: The watchdog destructor is not defined, define it using `watchdog.setDestructor()` ' +
+				'or `Watchdog.for` helper.',
+				null
+			);
+		}
+
+		this._elementOrData = elementOrData;
+
+		// Clone config because it might be shared within multiple watchdog instances. Otherwise
+		// when an error occurs in one of these editors the watchdog will restart all of them.
+		this._config = cloneDeepWith( config, function leaveDOMReferences( value ) {
+			return isElement( value ) ? value : undefined;
+		} );
+
+		return Promise.resolve()
+			.then( () => this._creator( elementOrData, this._config ) )
+			.then( editor => {
+				this._editor = editor;
+
+				window.addEventListener( 'error', this._boundErrorHandler );
+				this.listenTo( editor.model.document, 'change:data', this._throttledSave );
+
+				this._lastDocumentVersion = editor.model.document.version;
+				this._data = editor.data.get();
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Restarts the editor instance. This method is also called whenever an editor error occurs.
+	 * It fires the `restart` event.
+	 *
+	 * @fires restart
+	 * @returns {Promise.<module:core/editor/editor~Editor>}
+	 */
+	restart() {
+		this._throttledSave.flush();
+
+		return Promise.resolve()
+			.then( () => this.destroy() )
+			.catch( err => console.error( 'An error happened during the editor destructing.', err ) )
+			.then( () => {
+				if ( typeof this._elementOrData === 'string' ) {
+					return this.create( this._data, this._config );
+				}
+
+				const updatedConfig = Object.assign( {}, this._config, {
+					initialData: this._data
+				} );
+
+				return this.create( this._elementOrData, updatedConfig );
+			} )
+			.then( () => {
+				this.fire( 'restart' );
+			} );
+	}
+
+	/**
+	 * Destroys the current editor using the destructor passed to {@link #setDestructor} method.
+	 *
+	 * @returns {Promise.<module:watchdog/watchdog~Watchdog>}
+	 */
+	destroy() {
+		window.removeEventListener( 'error', this._boundErrorHandler );
+		this.stopListening( this._editor.model.document, 'change:data', this._throttledSave );
+
+		return Promise.resolve()
+			.then( () => this._destructor( this._editor ) )
+			.then( () => {
+				this._editor = null;
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Saves the editor data, so it can be restored after the crash even if the data cannot be fetched at
+	 * the moment of a crash.
+	 *
+	 * @private
+	 */
+	_save() {
+		const version = this._editor.model.document.version;
+
+		// Change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( version < this._lastDocumentVersion ) {
```

Comment:
When this condition is `true`? `this._lastDocumentVersion` is always set to current `document.version`, so it looks like `version` can be either same or bigger than `this._lastDocumentVersion`. Do I miss something?

Reply:
I'm not sure here. We created a similar thing in the `ckeditor5-autosave` in the past - https://github.com/ckeditor/ckeditor5-autosave/blob/master/src/autosave.js#L217-L228

cc @pjasiun.